Python/Product/AnacondaInstallLauncher/Program.cs

```diff
@@ -0,0 +1,212 @@
+            Console.Error.WriteLine("Starting:{0}  Filename: {1}{0}  Argument: {2}", Environment.NewLine, psi.FileName, psi.Arguments);
+            var p = Process.Start(psi);
+
+            p.OutputDataReceived += (_, e) => { if (e.Data != null) Console.Write(e.Data); };
+            p.ErrorDataReceived += (_, e) => { if (e.Data != null) Console.Error.Write(e.Data); };
+
+            return p;
+        }
+
+        int DoInstall() {
+            WaitForUninstall(false);
+            if (_cancel) {
+                return ERROR_INSTALL_CANCEL;
+            }
+
+            var p = Run(_installer, string.Format(CultureInfo.InvariantCulture, "/InstallationType=AllUsers /RegisterPython=0 /S /D={0}", _targetDir));
+            p.WaitForExit();
+            return p.ExitCode;
+        }
+
+        int DoUninstall() {
+            var p = Run(_installer, "/S");
+
+            if (p.WaitForExit(30000)) {
+                Console.Error.WriteLine("Process exited: {0}", p.ExitCode);
+            } else {
+                Console.Error.WriteLine("Process has not exited");
+            }
+
+            WaitForUninstall(true);
+
+            return _cancel ? ERROR_INSTALL_CANCEL : p.ExitCode;
+        }
+
+        void WaitForUninstall(bool requireWait) {
+            try {
+                WaitForProcess(DateTime.UtcNow.AddMinutes(30), requireWait);
+            } catch (Exception ex) {
+                Console.Error.WriteLine(ex.ToString());
+            }
+        }
+
+        bool KeepWaiting {
+            get {
+                if (_cancel) {
+                    return false;
+                }
+
+                if (!Directory.Exists(_targetDir)) {
+                    Console.Error.WriteLine("Target directory is not present");
+                    return false;
+                }
+                try {
+                    if (!Directory.EnumerateFileSystemEntries(_targetDir).Any()) {
+                        Console.Error.WriteLine("Target directory is empty");
+                        return false;
+                    }
+                } catch (Exception ex) {
+                    Console.Error.WriteLine("Error reading target directory");
+                    Console.Error.WriteLine(ex.ToString());
+                    return false;
+                }
+
+                return true;
+            }
+        }
+
+        Process[] GetProcesses() {
+            return Process.GetProcesses()
+                .Where(p => p.ProcessName.StartsWith("Un_") &&
+                    p.MainModule.FileVersionInfo.FileDescription.IndexOf("Anaconda", StringComparison.OrdinalIgnoreCase) >= 0)
+                .ToArray();
+        }
+
+        void WaitForProcess(DateTime stopAt, bool requireProcess) {
+            if (!KeepWaiting) {
+                return;
+            }
+
+            var procs = GetProcesses();
+            int retries = 30;
+            while (requireProcess && procs.Length == 0 && retries-- > 0) {
+                Thread.Sleep(1000);
```

Comment:
Why force context switching on the CPU every second when you could just wait on the process handle(s) (via `Process.WaitForExit`) like in the original example I provided?

Reply:
Because we're also checking the file system, and file system watchers are notoriously unreliable so I'd rather not use one just to avoid looping. This is not going to be the most CPU intensive operation while it's running (Windows Defender takes that prize...)

Reply:
You're not checking the file system in this loop. Waiting on the process handles then checking the file system would be less intensive.